lib/bridge/IrcHandler.js

```diff
@@ -705,15 +709,26 @@ IrcHandler.prototype.onPart = Promise.coroutine(function*(req, server, leavingUs
     }
 
     let promises = matrixRooms.map((room) => {
-        req.log.info("Leaving room %s", room.getId());
-        return this.ircBridge.getAppServiceBridge().getIntent(
-            matrixUser.getId()
-        ).leave(room.getId());
+        return this.delayWhileInflight(room, matrixUser).then(() => {
+            req.log.info("Leaving room %s", room.getId());
+            return this.ircBridge.getAppServiceBridge().getIntent(
+                matrixUser.getId()
+            ).leave(room.getId());
+        });
     });
     stats.membership(true, "part");
     yield Promise.all(promises);
 });
 
+IrcHandler.prototype.delayWhileInflight = function(room, matrixUser) {
+    if (this.ircBridge.userHasMessagesInFlight(room, matrixUser)) {
+        return Promise.delay(LEAVE_INFLIGHT_DELAY_MS).then(() => {
```

Comment:
There's a bit of a thundering herd problem here: Timers started at approximately the same time for the same duration are all going to terminate at the same time, leading to quite a lot of blocked operations unlocking all at once. Some jitter of ~100ms on each helps spread out the timer firing load, and therefore the potential for large blocking operations.